File: crypto/starkware/crypto/signature/signature.py
###############################################################################
# Copyright 2019 StarkWare Industries Ltd.                                    #
#                                                                             #
# Licensed under the Apache License, Version 2.0 (the "License").             #
# You may not use this file except in compliance with the License.            #
# You may obtain a copy of the License at                                     #
#                                                                             #
# https://www.starkware.co/open-source-license/                               #
#                                                                             #
# Unless required by applicable law or agreed to in writing,                  #
# software distributed under the License is distributed on an "AS IS" BASIS,  #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    #
# See the License for the specific language governing permissions             #
# and limitations under the License.                                          #
###############################################################################

import json
import math
import os
import random

from .math_utils import div_mod, ec_add, ec_double, ec_mult, is_quad_residue, sqrt_mod

PEDERSEN_HASH_POINT_FILENAME = os.path.join(
    os.path.dirname(__file__), 'pedersen_params.json')
PEDERSEN_PARAMS = json.load(open(PEDERSEN_HASH_POINT_FILENAME))

FIELD_PRIME = PEDERSEN_PARAMS['FIELD_PRIME']
FIELD_GEN = PEDERSEN_PARAMS['FIELD_GEN']
ALPHA = PEDERSEN_PARAMS['ALPHA']
BETA = PEDERSEN_PARAMS['BETA']
EC_ORDER = PEDERSEN_PARAMS['EC_ORDER']
CONSTANT_POINTS = PEDERSEN_PARAMS['CONSTANT_POINTS']

N_ELEMENT_BITS_ECDSA = math.floor(math.log(FIELD_PRIME, 2))
assert N_ELEMENT_BITS_ECDSA == 251

N_ELEMENT_BITS_HASH = math.ceil(math.log(FIELD_PRIME, 2))
assert N_ELEMENT_BITS_HASH == 252

# Elliptic curve parameters.
assert 2**N_ELEMENT_BITS_ECDSA < EC_ORDER < FIELD_PRIME

SHIFT_POINT = CONSTANT_POINTS[0]
MINUS_SHIFT_POINT = [SHIFT_POINT[0], FIELD_PRIME - SHIFT_POINT[1]]
EC_GEN = CONSTANT_POINTS[1]

assert SHIFT_POINT == [0x49ee3eba8c1600700ee1b87eb599f16716b0b1022947733551fde4050ca6804,
                       0x3ca0cfe4b3bc6ddf346d49d06ea0ed34e621062c0e056c1d0405d266e10268a]
assert EC_GEN == [0x1ef15c18599971b7beced415a40f0c7deacfd9b0d1819e03d723d8bc943cfca,
                  0x5668060aa49730b7be4801df46ec62de53ecd11abe43a32873000c36e8dc1f]


#########
# ECDSA #
#########


class InvalidPublicKeyError(Exception):
    def __init__(self):
        super().__init__('Given x coordinate does not represent any point on the elliptic curve.')


def get_y_coordinate(stark_key_x_coordinate):
    """
    Given the x coordinate of a stark_key, returns a possible y coordinate such that together the
    point (x,y) is on the curve.
    Note that the real y coordinate is either y or -y.
    If x is invalid stark_key it throws an error.
    """

    x = stark_key_x_coordinate
    y_squared = (x * x * x + ALPHA * x + BETA) % FIELD_PRIME
    if not is_quad_residue(y_squared, FIELD_PRIME):
        raise InvalidPublicKeyError()
    return sqrt_mod(y_squared, FIELD_PRIME)


def get_random_private_key():
    # NOTE: It is IMPORTANT to use a strong random function here.
    return random.randint(1, EC_ORDER - 1)


def private_key_to_ec_point_on_stark_curve(priv_key):
    assert 0 < priv_key < EC_ORDER
    return ec_mult(priv_key, EC_GEN, ALPHA, FIELD_PRIME)


def private_to_stark_key(priv_key):
    return private_key_to_ec_point_on_stark_curve(priv_key)[0]


def sign(msg_hash, priv_key):
    # Note: msg_hash must be smaller than 2**N_ELEMENT_BITS_ECDSA.
    # Message whose hash is >= 2**N_ELEMENT_BITS_ECDSA cannot be signed.
    # This happens with a very small probability.
    assert 0 <= msg_hash < 2**N_ELEMENT_BITS_ECDSA, 'Message not signable.'

    # Choose a valid k. In our version of ECDSA not every k value is valid,
    # and there is a negligible probability a drawn k cannot be used for signing.
    # This is why we have this loop.
    while True:
        # NOTE: It is IMPORTANT to use a strong random function here
        # or use a similar technique as in RFC6979
        # (pseudo-random defined by hash and secret-key).
        k = random.randint(1, EC_ORDER - 1)

        # Cannot fail because 0 < k < EC_ORDER and EC_ORDER is prime.
        x = ec_mult(k, EC_GEN, ALPHA, FIELD_PRIME)[0]

        # DIFF: in classic ECDSA, we take int(x) % n.
        r = int(x)
        if not (1 <= r < 2**N_ELEMENT_BITS_ECDSA):
            # Bad value. This fails with negligible probability.
            continue

        if (msg_hash + r * priv_key) % EC_ORDER == 0:
            # Bad value. This fails with negligible probability.
            continue

        w = div_mod(k, msg_hash + r * priv_key, EC_ORDER)
        if not (1 <= w < 2**N_ELEMENT_BITS_ECDSA):
            # Bad value. This fails with negligible probability.
            continue

        # DIFF: Here we send w instead of its inverse.
        return r, w


def mimic_ec_mult_air(m, point, shift_point):
    """
    Computes m * point + shift_point using the same steps like the AIR and throws an exception if
    and only if the AIR errors.
    """
    assert 0 < m < 2**N_ELEMENT_BITS_ECDSA
    partial_sum = shift_point
    for _ in range(N_ELEMENT_BITS_ECDSA):
        assert partial_sum[0] != point[0]
        if m & 1:
            partial_sum = ec_add(partial_sum, point, FIELD_PRIME)
        point = ec_double(point, ALPHA, FIELD_PRIME)
        m >>= 1
    assert m == 0
    return partial_sum


def verify(msg_hash, r, w, public_key):
    # Preassumptions:
    # DIFF: in classic ECDSA, we assert 1 <= r, w <= EC_ORDER-1.
    # Since r, w < 2**N_ELEMENT_BITS_ECDSA < EC_ORDER, we only need to verify r, w != 0.
    assert 1 <= r < 2**N_ELEMENT_BITS_ECDSA, 'r = %s' % r
    assert 1 <= w < 2**N_ELEMENT_BITS_ECDSA, 'w = %s' % w
    assert 0 <= msg_hash < 2**N_ELEMENT_BITS_ECDSA

    if isinstance(public_key, int):
        # Only the x coordinate of the point is given, check the two possibilities for the y
        # coordinate.
        try:
            y = get_y_coordinate(public_key)
        except InvalidPublicKeyError:
            return False
        assert pow(y, 2, FIELD_PRIME) == (
            pow(public_key, 3, FIELD_PRIME) + ALPHA * public_key + BETA) % FIELD_PRIME
        return verify(msg_hash, r, w, (public_key, y)) or \
            verify(msg_hash, r, w, (public_key, (-y) % FIELD_PRIME))
    else:
        # The public key is provided as a point.
        # Verify it is on the curve.
        assert (public_key[1]**2 - (public_key[0]**3 + ALPHA *
                                    public_key[0] + BETA)) % FIELD_PRIME == 0

    # Signature validation.
    # DIFF: original formula is:
    # x = (w*msg_hash)*EC_GEN + (w*r)*public_key
    # While what we implement is:
    # x = w*(msg_hash*EC_GEN + r*public_key).
    # While both mathematically equivalent, one might error while the other doesn't,
    # given the current implementation.
    # This formula ensures that if the verification errors in our AIR, it errors here as well.
    try:
        zG = mimic_ec_mult_air(msg_hash, EC_GEN, MINUS_SHIFT_POINT)
        rQ = mimic_ec_mult_air(r, public_key, SHIFT_POINT)
        wB = mimic_ec_mult_air(w, ec_add(zG, rQ, FIELD_PRIME), SHIFT_POINT)
        x = ec_add(wB, MINUS_SHIFT_POINT, FIELD_PRIME)[0]
    except AssertionError:
        return False

    # DIFF: Here we drop the mod n from classic ECDSA.
    return r == x


#################
# Pedersen hash #
#################

def pedersen_hash(*elements):
    return pedersen_hash_as_point(*elements)[0]


def pedersen_hash_as_point(*elements):
    """
    Similar to pedersen_hash but also returns the y coordinate of the resulting EC point.
    This function is used for testing.
    """
    point = SHIFT_POINT
    for i, x in enumerate(elements):
        assert 0 <= x < FIELD_PRIME
        point_list = CONSTANT_POINTS[2 + i * N_ELEMENT_BITS_HASH:2 + (i + 1) * N_ELEMENT_BITS_HASH]
        assert len(point_list) == N_ELEMENT_BITS_HASH
        for pt in point_list:
            assert point[0] != pt[0], 'Unhashable input.'
            if x & 1:
                point = ec_add(point, pt, FIELD_PRIME)
            x >>= 1
        assert x == 0
    return point

#############################
# party_a/party_b signature #
#############################


def get_msg(instruction_type, vault0, vault1, amount0, amount1, token0,
            token1_or_pub_key, nonce, expiration_timestamp, hash=pedersen_hash):
    """
    Creates a message to sign on.
    """
    packed_message = instruction_type
    packed_message = packed_message * 2**31 + vault0
    packed_message = packed_message * 2**31 + vault1
    packed_message = packed_message * 2**63 + amount0
    packed_message = packed_message * 2**63 + amount1
    packed_message = packed_message * 2**31 + nonce
    packed_message = packed_message * 2**22 + expiration_timestamp
    return hash(hash(token0, token1_or_pub_key), packed_message)


def get_limit_order_msg(vault_sell, vault_buy, amount_sell, amount_buy, token_sell,
                        token_buy, nonce, expiration_timestamp, hash=pedersen_hash):
    """
    party_a sells amount_sell coins of token_sell from vault_sell.
    party_a buys amount_buy coins of token_buy into vault_buy.
    """
    assert 0 <= vault_sell < 2**31
    assert 0 <= vault_buy < 2**31
    assert 0 <= amount_sell < 2**63
    assert 0 <= amount_buy < 2**63
    assert 0 <= token_sell < FIELD_PRIME
    assert 0 <= token_buy < FIELD_PRIME
    assert 0 <= nonce < 2**31
    assert 0 <= expiration_timestamp < 2**22

    instruction_type = 0
    return get_msg(instruction_type, vault_sell, vault_buy, amount_sell, amount_buy, token_sell,
                   token_buy, nonce, expiration_timestamp, hash=hash)


def get_transfer_msg(amount, nonce, sender_vault_id, token, receiver_vault_id,
                     receiver_public_key, expiration_timestamp, hash=pedersen_hash):
    """
    Transfer `amount` of `token` from `sender_vault_id` to `receiver_vault_id`.
    """
    assert 0 <= sender_vault_id < 2**31
    assert 0 <= receiver_vault_id < 2**31
    assert 0 <= amount < 2**63
    assert 0 <= token < FIELD_PRIME
    assert 0 <= receiver_public_key < FIELD_PRIME
    assert 0 <= nonce < 2**31
    assert 0 <= expiration_timestamp < 2**22

    instruction_type = 1
    return get_msg(instruction_type, sender_vault_id, receiver_vault_id, amount, 0, token,
                   receiver_public_key, nonce, expiration_timestamp, hash=hash)

File: crypto/starkware/crypto/signature/signature.js
/////////////////////////////////////////////////////////////////////////////////
// Copyright 2019 StarkWare Industries Ltd.                                    //
//                                                                             //
// Licensed under the Apache License, Version 2.0 (the "License").             //
// You may not use this file except in compliance with the License.            //
// You may obtain a copy of the License at                                     //
//                                                                             //
// https://www.starkware.co/open-source-license/                               //
//                                                                             //
// Unless required by applicable law or agreed to in writing,                  //
// software distributed under the License is distributed on an "AS IS" BASIS,  //
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.    //
// See the License for the specific language governing permissions             //
// and limitations under the License.                                          //
/////////////////////////////////////////////////////////////////////////////////

const BN = require('bn.js');
const hash = require('hash.js');
const { curves: eCurves, ec: EllipticCurve } = require('elliptic');
const assert = require('assert');
const constantPointsHex = require('./constant_points.json');

// Equals 2**251 + 17 * 2**192 + 1.
const prime = new BN('800000000000011000000000000000000000000000000000000000000000001', 16);
// Equals 2**251. This value limits msgHash and the signature parts.
const maxEcdsaVal =
    new BN('800000000000000000000000000000000000000000000000000000000000000', 16);

// Generate BN of used constants.
const zeroBn = new BN('0', 16);
const oneBn = new BN('1', 16);
const twoBn = new BN('2', 16);
const twoPow22Bn = new BN('400000', 16);
const twoPow31Bn = new BN('80000000', 16);
const twoPow63Bn = new BN('8000000000000000', 16);

// Create a curve with stark curve parameters.
const starkEc = new EllipticCurve(
    new eCurves.PresetCurve({
        type: 'short',
        prime: null,
        p: prime,
        a: '00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001',
        b: '06f21413 efbe40de 150e596d 72f7a8c5 609ad26c 15c915c1 f4cdfcb9 9cee9e89',
        n: '08000000 00000010 ffffffff ffffffff b781126d cae7b232 1e66a241 adc64d2f',
        hash: hash.sha256,
        gRed: false,
        g: constantPointsHex[1]
    })
);

const constantPoints = constantPointsHex.map(coords => (
    starkEc.curve.point(new BN(coords[0], 16), new BN(coords[1], 16))));
const shiftPoint = constantPoints[0];

/*
  Checks that the string str start with '0x'.
*/
function hasHexPrefix(str) {
    return str.substring(0, 2) === '0x';
}

/*
 Asserts input is equal to or greater then lowerBound and lower then upperBound.
 Assert message specifies inputName.
 input, lowerBound, and upperBound should be of type BN.
 inputName should be a string.
*/
function assertInRange(
        input,
        lowerBound,
        upperBound,
        inputName = ''
) {
    const messageSuffix = inputName === '' ? 'invalid length' : `invalid ${inputName} length`;
    assert(
        input.gte(lowerBound) && input.lt(upperBound), `Message not signable, ${messageSuffix}.`
    );
}

/*
 Full specification of the hash function can be found here:
   https://starkware.co/starkex/docs/signatures.html#pedersen-hash-function
 shiftPoint was added for technical reasons to make sure the zero point on the elliptic curve does
 not appear during the computation. constantPoints are multiples by powers of 2 of the constant
 points defined in the documentation.
*/
function pedersen(input) {
    let point = shiftPoint;
    for (let i = 0; i < input.length; i++) {
        let x = new BN(input[i], 16);
        assert(x.gte(zeroBn) && x.lt(prime), 'Invalid input: ' + input[i]);
        for (let j = 0; j < 252; j++) {
            const pt = constantPoints[2 + i * 252 + j];
            assert(!point.getX().eq(pt.getX()));
            if (x.and(oneBn).toNumber() !== 0) {
                point = point.add(pt);
            }
            x = x.shrn(1);
        }
    }
    return point.getX().toString(16);
}

function hashMsg(
        instructionTypeBn,
        vault0Bn,
        vault1Bn,
        amount0Bn,
        amount1Bn,
        nonceBn,
        expirationTimestampBn,
        token0,
        token1OrPubKey,
        condition = null
) {
    let packedMessage = instructionTypeBn;
    packedMessage = packedMessage.ushln(31).add(vault0Bn);
    packedMessage = packedMessage.ushln(31).add(vault1Bn);
    packedMessage = packedMessage.ushln(63).add(amount0Bn);
    packedMessage = packedMessage.ushln(63).add(amount1Bn);
    packedMessage = packedMessage.ushln(31).add(nonceBn);
    packedMessage = packedMessage.ushln(22).add(expirationTimestampBn);
    let msgHash = null;
    if (condition === null) {
        msgHash = pedersen([
            pedersen([token0, token1OrPubKey]), packedMessage.toString(16)
        ]);
    } else {
        msgHash = pedersen([
            pedersen([
                pedersen([token0, token1OrPubKey]), condition
            ]), packedMessage.toString(16)
        ]);
    }

    const msgHashBN = new BN(msgHash, 16);
    assertInRange(msgHashBN, zeroBn, maxEcdsaVal, 'msgHash');
    return msgHash;
}

/*
 Serializes the order message in the canonical format expected by the verifier.
 party_a sells amountSell coins of tokenSell from vaultSell.
 party_a buys amountBuy coins of tokenBuy into vaultBuy.

 Expected types:
 ---------------
 vaultSell, vaultBuy - uint31 (as int)
 amountSell, amountBuy - uint63 (as decimal string)
 tokenSell, tokenBuy - uint256 field element strictly less than the prime (as hex string with 0x)
 nonce - uint31 (as int)
 expirationTimestamp - uint22 (as int).
*/
function getLimitOrderMsgHash(
        vaultSell,
        vaultBuy,
        amountSell,
        amountBuy,
        tokenSell,
        tokenBuy,
        nonce,
        expirationTimestamp
) {
    assert(
        hasHexPrefix(tokenSell) && hasHexPrefix(tokenBuy),
        'Hex strings expected to be prefixed with 0x.'
    );
    const vaultSellBn = new BN(vaultSell);
    const vaultBuyBn = new BN(vaultBuy);
    const amountSellBn = new BN(amountSell, 10);
    const amountBuyBn = new BN(amountBuy, 10);
    const tokenSellBn = new BN(tokenSell.substring(2), 16);
    const tokenBuyBn = new BN(tokenBuy.substring(2), 16);
    const nonceBn = new BN(nonce);
    const expirationTimestampBn = new BN(expirationTimestamp);

    assertInRange(vaultSellBn, zeroBn, twoPow31Bn);
    assertInRange(vaultBuyBn, zeroBn, twoPow31Bn);
    assertInRange(amountSellBn, zeroBn, twoPow63Bn);
    assertInRange(amountBuyBn, zeroBn, twoPow63Bn);
    assertInRange(tokenSellBn, zeroBn, prime);
    assertInRange(tokenBuyBn, zeroBn, prime);
    assertInRange(nonceBn, zeroBn, twoPow31Bn);
    assertInRange(expirationTimestampBn, zeroBn, twoPow22Bn);

    const instructionType = zeroBn;
    return hashMsg(
        instructionType,
        vaultSellBn,
        vaultBuyBn,
        amountSellBn,
        amountBuyBn,
        nonceBn,
        expirationTimestampBn,
        tokenSell.substring(2),
        tokenBuy.substring(2)
    );
}

/*
 Serializes the transfer message in the canonical format expected by the verifier.
 The sender transfer 'amount' coins of 'token' from vault with id senderVaultId to vault with id
 receiverVaultId. The receiver's public key is receiverPublicKey.
 If a condition is added, it is verified before executing the transfer. The format of the condition
 is defined by the application.
 Expected types:
 ---------------
 amount - uint63 (as decimal string)
 nonce - uint31 (as int)
 senderVaultId uint31 (as int)
 token - uint256 field element strictly less than the prime (as hex string with 0x)
 receiverVaultId - uint31 (as int)
 receiverPublicKey - uint256 field element strictly less than the prime (as hex string with 0x)
 expirationTimestamp - uint22 (as int).
 condition - uint256 field element strictly less than the prime (as hex string with 0x)
*/
function getTransferMsgHash(
        amount,
        nonce,
        senderVaultId,
        token,
        receiverVaultId,
        receiverPublicKey,
        expirationTimestamp,
        condition = null
) {
    assert(
        hasHexPrefix(token) && hasHexPrefix(receiverPublicKey) && (condition === null ||
            hasHexPrefix(condition)), 'Hex strings expected to be prefixed with 0x.'
    );
    const amountBn = new BN(amount, 10);
    const nonceBn = new BN(nonce);
    const senderVaultIdBn = new BN(senderVaultId);
    const tokenBn = new BN(token.substring(2), 16);
    const receiverVaultIdBn = new BN(receiverVaultId);
    const receiverPublicKeyBn = new BN(receiverPublicKey.substring(2), 16);
    const expirationTimestampBn = new BN(expirationTimestamp);

    assertInRange(amountBn, zeroBn, twoPow63Bn);
    assertInRange(nonceBn, zeroBn, twoPow31Bn);
    assertInRange(senderVaultIdBn, zeroBn, twoPow31Bn);
    assertInRange(tokenBn, zeroBn, prime);
    assertInRange(receiverVaultIdBn, zeroBn, twoPow31Bn);
    assertInRange(receiverPublicKeyBn, zeroBn, prime);
    assertInRange(expirationTimestampBn, zeroBn, twoPow22Bn);
    let instructionType = oneBn;
    if (condition !== null) {
        condition = condition.substring(2);
        assertInRange(new BN(condition), zeroBn, prime, 'condition');
        instructionType = twoBn;
    }
    return hashMsg(
        instructionType,
        senderVaultIdBn,
        receiverVaultIdBn,
        amountBn,
        zeroBn,
        nonceBn,
        expirationTimestampBn,
        token.substring(2),
        receiverPublicKey.substring(2),
        condition
    );
}

/*
 The function _truncateToN in lib/elliptic/ec/index.js does a shift-right of delta bits,
 if delta is positive, where
   delta = msgHash.byteLength() * 8 - starkEx.n.bitLength().
 This function does the opposite operation so that
   _truncateToN(fixMsgHashLen(msgHash)) == msgHash.
*/
function fixMsgHashLen(msgHash) {
    // Convert to BN to remove leading zeros.
    msgHash = new BN(msgHash, 16).toString(16);

    if (msgHash.length <= 62) {
        // In this case, msgHash should not be transformed, as the byteLength() is at most 31,
        // so delta < 0 (see _truncateToN).
        return msgHash;
    }
    assert(msgHash.length === 63);
    // In this case delta will be 4 so we perform a shift-left of 4 bits by adding a zero.
    return msgHash + '0';
}

/*
 Signs a message using the provided key.
 privateKey should be an elliptic.keyPair with a valid private key.
 Returns an elliptic.Signature.
*/
function sign(privateKey, msgHash) {
    const msgHashBN = new BN(msgHash, 16);
    // Verify message hash has valid length.
    assertInRange(msgHashBN, zeroBn, maxEcdsaVal, 'msgHash');
    const msgSignature = privateKey.sign(fixMsgHashLen(msgHash));
    const { r, s } = msgSignature;
    const w = s.invm(starkEc.n);
    // Verify signature has valid length.
    assertInRange(r, oneBn, maxEcdsaVal, 'r');
    assertInRange(s, oneBn, starkEc.n, 's');
    assertInRange(w, oneBn, maxEcdsaVal, 'w');
    return msgSignature;
}

/*
 Verifies a message using the provided key.
 publicKey should be an elliptic.keyPair with a valid public key.
 msgSignature should be an elliptic.Signature.
 Returns a boolean true if the verification succeeds.
*/
function verify(publicKey, msgHash, msgSignature) {
    const msgHashBN = new BN(msgHash, 16);
    // Verify message hash has valid length.
    assertInRange(msgHashBN, zeroBn, maxEcdsaVal, 'msgHash');
    const { r, s } = msgSignature;
    const w = s.invm(starkEc.n);
    // Verify signature has valid length.
    assertInRange(r, oneBn, maxEcdsaVal, 'r');
    assertInRange(s, oneBn, starkEc.n, 's');
    assertInRange(w, oneBn, maxEcdsaVal, 'w');
    return publicKey.verify(fixMsgHashLen(msgHash), msgSignature);
}

module.exports = {
    prime, ec: starkEc, constantPoints, shiftPoint, maxEcdsaVal,  // Data.
    pedersen, getLimitOrderMsgHash, getTransferMsgHash, sign, verify  // Function.
};
